Sources/MongoSwift/MongoConnectionString.swift

```diff
@@ -0,0 +1,193 @@
+import Foundation
+
+/// Represents a MongoDB connection string.
+/// - SeeAlso: https://docs.mongodb.com/manual/reference/connection-string/
+public struct MongoConnectionString: Codable, LosslessStringConvertible {
+    /// Represents a connection string scheme.
+    public struct Scheme: LosslessStringConvertible, Equatable {
+        /// Indicates that this connection string uses the scheme `mongodb`.
+        public static let mongodb = Scheme(.mongodb)
+
+        /// Indicates that this connection string uses the scheme `mongodb+srv`.
+        public static let srv = Scheme(.srv)
+
+        /// Internal representation of a scheme.
+        private enum _Scheme: String {
+            case mongodb
+            case srv = "mongodb+srv"
+        }
+
+        private let _scheme: _Scheme
+
+        private init(_ value: _Scheme) {
+            self._scheme = value
+        }
+
+        /// LosslessStringConvertible` protocol requirements
+        public init?(_ description: String) {
+            guard let _scheme = _Scheme(rawValue: description) else {
+                return nil
+            }
+            self.init(_scheme)
+        }
+
+        public var description: String { self._scheme.rawValue }
+    }
+
+    /// A struct representing a host identifier, consisting of a host and an optional port.
+    /// In standard connection strings, this describes the address of a mongod or mongos to connect to.
+    /// In mongodb+srv connection strings, this describes a DNS name to be queried for SRV and TXT records.
+    public struct HostIdentifier: Equatable, CustomStringConvertible {
+        private static func parsePort(from: String) throws -> UInt16? {
+            guard let port = UInt16(from), port > 0 else {
+                throw MongoError.InvalidArgumentError(
+                    message: "port must be a valid, positive unsigned 16 bit integer"
+                )
+            }
+            return port
+        }
+
+        private enum HostType: String {
+            case ipv4
+            case ipLiteral = "ip_literal"
+            case hostname
+            case unixDomainSocket
+        }
+
+        /// The hostname or IP address.
+        public let host: String
+
+        /// The port number.
+        public let port: UInt16?
+
+        private let type: HostType
+
+        /// Initializes a ServerAddress, using the default localhost:27017 if a host/port is not provided.
+        internal init(_ hostAndPort: String = "localhost:27017") throws {
+            guard !hostAndPort.contains("?") else {
+                throw MongoError.InvalidArgumentError(message: "\(hostAndPort) contains invalid characters")
+            }
+
+            var host: String
```

Comment:
do we need these two variables? Or can we just assign to `self.host` and `self.port` directly?

Reply:
we can assign it directly